### ISSUE TITLE
Consider implementing partially-applied versions of RQL terms

### ISSUE THREAD

**Issue by marshall007** [neumino/rethinkdbdash#59]
There has been a lot of discussion around implementing curried/partially applied terms in the core driver (see [#1190](https://github.com/rethinkdb/rethinkdb/issues/1190), [#2578](https://github.com/rethinkdb/rethinkdb/issues/2578#issuecomment-69282165), and [#3485](https://github.com/rethinkdb/rethinkdb/issues/3485)). From what I've gathered this will be difficult/impossible to implement consistently across all drivers. For example in the Ruby driver, all terms can be called either infix or prefix so you can't determine whether a term should be evaluated or wait on additional arguments.

That being said it does seem quite do-able in the JS driver (see my comments on [#2578](https://github.com/rethinkdb/rethinkdb/issues/2578#issuecomment-69282165)). So, do we need to maintain 1:1 parity with the core driver? If so would you consider implementing some kind of plugin system that allows us to override/extend RQL terms?

**Comment by marshall007** [comment]
Some context on the kind of operations this would allow you to do:

```js
r.expr(['hello']).contains(r.match('^h')) // -> true
r.expr([1]).map(r.add(1)) // -> [2]
r.expr([false]).map(r.or(true)) // -> [true]
```

And an example implementation from the other thread (we would obviously do this in a more generic way if implemented):

```js
r.contains = function () {
  var args = arguments;
  return function (row) {
    return row.contains.apply(row, args);
  }
};

r.expr([[1], [2]]).filter(r.contains(1)) // -> [ [1] ]
```

**Comment by neumino** [comment]
So I have to think a bit more about how to do this, but `thinky` which uses rethinkdbdash can extends things. You can do stuff like
```
Users.define("isAdult", function(doc) {
    return doc("age").gt(18)
});

Model.isAdult().run().then(function(result) {
    // ...
})
// and under the hood, Model is just `r.table("users")`
```

Thinky wraps queries with its own class (and dynamically import all the methods). It should be possible to push this functionality lower (in rethinkdbdash).
That being say, you probably won't be able to overwrite terms (like changing the behavior of map - you would have to use something like `mapWith`).

As for the infix vs prefix, nothing really prevents the driver for supporting both. The only slightly annoying things (with the current implementation) is how to represent the backtrace of a broken query. Like what syntax to pick between these two:

```
r.table([1,2,3]).map(r.row.add(2))
r.table([1,2,3]).map(r.add(r.row, 2))
```

I wrote "with the current implementation" because I think we could work around this by having two "different" methods that points to the same ReQL methods.

So TL;DR: I think it's possible, I need to think a bit about it on how to properly do it.

**Comment by neumino** [comment]
Also, thanks for the suggestion @marshall007 :)

**Comment by neumino** [comment]
So some time ago I exposed `Term` so now it's actually already possible to implement terms.

```
var r = require('rethinkdbdash');
r.prototype.contains = function(value) {
  // ...
}

r = r();
var Term = r._Term;
Term.prototype.isAdult = function() {
  // ...
}
```
